pyteal/ast/subroutine.py

```diff
@@ -446,53 +620,93 @@ def evaluateSubroutine(subroutine: SubroutineDefinition) -> SubroutineDeclaratio
     Type 1 (by-value): these have python type Expr
     Type 2 (by-reference): these have python type ScratchVar
     Type 3 (ABI): these are ABI typed variables with scratch space storage, and still pass by value
+    Type 4 (ABI-output-arg): ABI typed variables with scratch space, a new ABI instance is generated inside function body,
+        not one of the cases in the previous three options
 
     Usage (A) "argumentVars" - Storing pre-placed stack variables into local scratch space:
         Type 1. (by-value) use ScratchVar.store() to pick the actual value into a local scratch space
         Type 2. (by-reference) ALSO use ScratchVar.store() to pick up from the stack
             NOTE: SubroutineCall.__teal__() has placed the _SLOT INDEX_ on the stack so this is stored into the local scratch space
         Type 3. (ABI) abi_value.stored_value.store() to pick from the stack
+        Type 4. (ABI-output-arg) it is not really used here, since it is only generated internal of the subroutine
 
     Usage (B) "loadedArgs" - Passing through to an invoked PyTEAL subroutine AST:
         Type 1. (by-value) use ScratchVar.load() to have an Expr that can be compiled in python by the PyTEAL subroutine
         Type 2. (by-reference) use a DynamicScratchVar as the user will have written the PyTEAL in a way that satisfies
             the ScratchVar API. I.e., the user will write `x.load()` and `x.store(val)` as opposed to just `x`.
         Type 3. (ABI) use abi_value itself after storing stack value into scratch space.
+        Type 4. (ABI-output-arg) generates a new instance of the ABI value,
+            and appends a return expression of stored value of the ABI keyword value.
     """
 
     def var_n_loaded(
         param: str,
-    ) -> Tuple[ScratchVar, Union[ScratchVar, abi.BaseType, Expr]]:
-        loaded: Union[ScratchVar, abi.BaseType, Expr]
-        argVar: ScratchVar
+    ) -> tuple[ScratchVar, ScratchVar | abi.BaseType | Expr]:
+        _loaded_var: ScratchVar | abi.BaseType | Expr
+        _argument_var: ScratchVar
 
         if param in subroutine.by_ref_args:
-            argVar = DynamicScratchVar(TealType.anytype)
-            loaded = argVar
+            _argument_var = DynamicScratchVar(TealType.anytype)
+            _loaded_var = _argument_var
         elif param in subroutine.abi_args:
             internal_abi_var = subroutine.abi_args[param].new_instance()
```

Comment:
`internal_abi_var` is unsafe because it only exists for the abi case, but there could be other cases that unwittingly attempt to access it.